torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -595,22 +659,82 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupGloo::scatter(
   throw std::runtime_error("ProcessGroupGloo does not support scatter");
 }
 
+at::Tensor& checkSingleTensor(std::vector<at::Tensor>& tensors) {
+  if (tensors.size() != 1) {
+    throw std::runtime_error("ProcessGroupGloo::send takes a single tensor");
+  }
+  auto& tensor = tensors[0];
+  if (!tensor.is_contiguous()) {
+    throw std::runtime_error("input tensor has to be contiguous");
+  }
+  if (tensor.is_sparse()) {
+    throw std::runtime_error("input tensor has to be dense");
+  }
+  return tensor;
+}
+
 std::shared_ptr<ProcessGroup::Work> ProcessGroupGloo::send(
-    std::vector<at::Tensor>& /* unused */,
-    int /* unused */) {
-  throw std::runtime_error("ProcessGroupGloo does not support send");
+    std::vector<at::Tensor>& tensors,
+    int dstRank) {
+  auto& tensor = checkSingleTensor(tensors);
+  auto ptr = tensor.data_ptr();
+  auto size = tensor.numel() * tensor.type().elementSizeInBytes();
+
+  // Construct unbound buffer.
+  auto& context = contexts_[0];
+  auto buf = context->createUnboundBuffer(ptr, size);
+  buf->send(dstRank, 0);
+
+  // The work captures the tensor to prevent it being deallocated and
+  // the unbound buffer to synchronize on completion of the send.
+  return std::shared_ptr<ProcessGroup::Work>(
```

Comment:
return make_shared is better, in case new fails and throws

Reply:
I'll see if this works -- it returns a base class but has to instantiate the sub class.